test/functional/placeholder.js

```diff
@@ -0,0 +1,94 @@
+/**
+ * Copyright 2015 The Incremental DOM Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+var IncrementalDOM = require('../../index'),
+    patch = IncrementalDOM.patch,
+    elementOpen = IncrementalDOM.elementOpen,
+    elementClose = IncrementalDOM.elementClose,
+    elementVoid = IncrementalDOM.elementVoid,
+    elementPlaceholder = IncrementalDOM.elementPlaceholder;
+
+describe('placeholders', () => {
+  var container;
+
+  beforeEach(() => {
+    container = document.createElement('div');
+    document.body.appendChild(container);
+  });
+
+  afterEach(() => {
+    document.body.removeChild(container);
+  });
+
+  describe('without an existing DOM element', () => {
+    var el;
+
+    beforeEach(() => {
+      patch(container, () => {
+        el = elementPlaceholder('div', 'placeholder', ['class', 'someClass'], 'data-foo', 'bar');
+      });
+    });
+
+    it('should render with the specified tag', () => {
+      expect(el.tagName).to.equal('DIV');
+    });
+
+    it('should render with static attributes', () => {
+      expect(el.className).to.equal('someClass');
+    });
+
+    it('should not render with dynamic attributes', () => {
+      expect(el.getAttribute('data-foo')).not.to.equal('bar');
+    });
+  });
+
+  describe('with an existing DOM element', () => {
+    var div;
+    var el;
+    var child;
+
+    beforeEach(() => {
+      patch(container, function() {
+        div = elementPlaceholder('div', 'placeholder');
+        patch(div, function() {
+          child = elementVoid('div', null, ['id', 'child']);
+        });
+      });
+
+      patch(container, () => {
+        el = elementPlaceholder('span', 'placeholder', ['class', 'someClass'], 'data-foo', 'bar');
+      });
+    });
+
+    it('should keep the element', () => {
+      expect(el).to.equal(div);
+      expect(div.tagName).to.equal('DIV');
+    });
+
+    it('should not set static attributes', () => {
+      expect(div.getAttribute('class')).not.to.equal('someClass');
+    });
+
+    it('should not set dynamic attributes', () => {
+      expect(div.getAttribute('data-foo')).not.to.equal('bar');
```

Comment:
>  The outside should be controlled by the one instantiating the component with the inside controlled by the component itself.

The pattern of the outside controlling the container while the render function controls everything inside it doesn't quite apply anymore with the [recommendation](https://github.com/google/incremental-dom/pull/66#issuecomment-121853982) to pull out an `updateAttributes` function from general DOM alignment. 

You can't do a normal `elementOpen`/`alignWithDOM` on the `container` because then your container reference could point to a removed element. But `updateAttributes` would only mutate the attributes on the element, never removing it. Now we can treat the container as just a root element, and the view/component/instance attached to that root element is able to exist in isolation from a parent. That principle is one of the things that makes Backbone so damn nice: view's don't require a parent to exist and are fully autonomous and useful without one.

> For example, you will see people passing classes to components, which will conditionally add it to a class that they themselves define. It ends up that many components do this, but you never know which ones do it and which ones don't without going into the component itself and checking. 

You'd still be able to mutate a DOM node outside of the normal flow with `updateAttributes`, even inside the same `render`.

But I think breaking encapsulation by supporting this as a first class API is a much more harmful. Imagine some subview needs a `special` class for the correct CSS styles. But now, a parent comes along and tells the subview to remove it's class.